### PR TITLE
Core: Replace `lodash` with `es-toolkit`

### LINKED ISSUE #28611: [Tracking]: Replace Lodash Usage with Native Functionality or Alternatives
## Overview

In our efforts to optimize our monorepo's dependencies and improve bundle sizes, we are looking to replace our usage of Lodash with native JavaScript functionality or alternative libraries where applicable. This initiative is inspired by the recommendations found in [You Don't Need Lodash/Underscore](https://github.com/you-dont-need/You-Dont-Need-Lodash-Underscore), which outlines many cases where modern JavaScript provides solutions that previously required utility libraries like Lodash.

## Specific Lodash Functions in Use

From an initial audit, the following Lodash functions are currently in use across various packages in our monorepo:

- [ ] `lodash/camelCase`
- [ ] `lodash/cloneDeep`
- [ ] `lodash/countBy`
- [ ] `lodash/debounce`
- [ ] `lodash/isEqual`
- [ ] `lodash/isPlainObject`
- [ ] `lodash/kebabCase`
- [ ] `lodash/mapKeys`
- [ ] `lodash/mapValues`
- [ ] `lodash/merge`
- [ ] `lodash/mergeWith`
- [ ] `lodash/pick`
- [ ] `lodash/pickBy`
- [ ] `lodash/startCase`
- [ ] `lodash/throttle`
- [ ] `lodash/uniq`
- [ ] `lodash/upperFirst`

## Acceptance Criteria

To consider this work complete, the following criteria must be met:

1. **No Direct Lodash Imports:** Direct imports from Lodash (e.g., `import merge from 'lodash/merge'`) should be replaced across the monorepo.
2. **Functionality Preservation:** Replacements must maintain existing functionality. Unit tests should pass without modifications to the tests themselves (except of changing the way how lodash functionalities might get mocked), indicating that behavior is preserved.
3. **Bundle Size Reduction:** The overall bundle size of our packages should decrease as a result of removing Lodash, measured by our current bundle analysis tools.
4. **Performance Considerations:** Any native or alternative solutions should not introduce significant performance regressions.
5. **Code Review:** Changes should be reviewed for readability and maintainability. While functional replacements are the goal, we should avoid overly complex or cryptic solutions in favor of clarity.

## Potential Challenges

- **Complexity of `merge` and `mergeWith`:** These functions are used extensively for deep object merging. Identifying a suitable replacement that handles all edge cases might be challenging.
- **Throttling:** The `lodash/throttle` function is used in performance-sensitive areas. Ensuring that the replacement maintains the same throttling behavior and performance characteristics is crucial.

## Request for Contributions

We are seeking contributions from the open source community to tackle this initiative. Contributions can range from identifying and documenting potential replacements, to implementing changes in specific packages. Please comment on this issue if you're able to take on a part of this work, specifying which packages or Lodash functions you're focusing on.

Thank you for your contributions to improving our monorepo's efficiency and maintainability!

### PR DESCRIPTION
**Pull request by ndelangen**
Closes: https://github.com/storybookjs/storybook/issues/28611

## What I did

`lodash` is hard to tree-shake, it's CJS, and we want to move to less heavy, ESM dependencies.

This replaces `lodash` for `es-toolkit` which is mostly a drop-in replacement :hearts:

I've only changed the core for now.

**before**:
Screenshot 2024-08-27 at 16 02 20

**after**:
Screenshot 2024-08-27 at 16 02 26

#### The changes in this PR are covered in the following automated tests:
- [ ] stories
- [ ] unit tests
- [ ] integration tests
- [ ] end-to-end tests

#### Manual testing

_This section is mandatory for all contributions. If you believe no manual test is necessary, please state so explicitly. Thanks!_

### Documentation

- [ ] Add or update documentation reflecting your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

## Checklist for Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below:
   <details>
     <summary>Available labels</summary>

     - `bug`: Internal changes that fixes incorrect behavior.
     - `maintenance`: User-facing maintenance tasks.
     - `dependencies`: Upgrading (sometimes downgrading) dependencies.
     - `build`: Internal-facing build tooling & test updates. Will not show up in release changelog.
     - `cleanup`: Minor cleanup style change. Will not show up in release changelog.
     - `documentation`: Documentation **only** changes. Will not show up in release changelog.
     - `feature request`: Introducing a new feature.
     - `BREAKING CHANGE`: Changes that break compatibility in some way with current major version.
     - `other`: Changes that don't fit in the above categories.
   
   </details>

### 🦋 Canary release

This PR does not have a canary release associated. You can request a canary release of this pull request by mentioning the `@storybookjs/core` team here.

_core team members can create a canary release [here](https://github.com/storybookjs/storybook/actions/workflows/canary-release-pr.yml) or locally with `gh workflow run --repo storybookjs/storybook canary-release-pr.yml --field pr=<PR_NUMBER>`_

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createSize |  0 B | 0 B | 0 B | - | - |
| generateSize |  77.5 MB | 77.5 MB | 0 B | 0.91 | 0% |
| initSize |  162 MB | 163 MB | 661 kB | **3.29** | 0.4% |
| diffSize |  84.6 MB | 85.3 MB | 661 kB | **2.7** | 0.8% |
| buildSize |  7.19 MB | 7.34 MB | 154 kB | **-1.64** | 2.1% |
| buildSbAddonsSize |  1.66 MB | 1.57 MB | -95.5 kB | **-281.8** | 🔰-6.1% |
| buildSbCommonSize |  195 kB | 195 kB | 0 B | - | 0% |
| buildSbManagerSize |  1.96 MB | 2.3 MB | 336 kB | -0.06 | 14.6% |
| buildSbPreviewSize |  352 kB | 311 kB | -40.8 kB | **-Infinity** | 🔰-13.1% |
| buildStaticSize |  0 B | 0 B | 0 B | - | - |
| buildPrebuildSize |  4.17 MB | 4.37 MB | 200 kB | **-1.25** | 4.6% |
| buildPreviewSize |  3.02 MB | 2.97 MB | -45.7 kB | **-128.58** | -1.5% |
| testBuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbAddonsSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbCommonSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbManagerSize |  0 B | 0 B | 0 B | - | - |
| testBuildSbPreviewSize |  0 B | 0 B | 0 B | - | - |
| testBuildStaticSize |  0 B | 0 B | 0 B | - | - |
| testBuildPrebuildSize |  0 B | 0 B | 0 B | - | - |
| testBuildPreviewSize |  0 B | 0 B | 0 B | - | - |

| name | before | after | diff | z | % |
| ---- | ------ | ----- | ---- | - | - |
| createTime |  5.7s | 6s | 312ms | -0.9 | 5.1% |
| generateTime |  19.8s | 23s | 3.1s | 0.21 | 13.8% |
| initTime |  14.3s | 16.9s | 2.5s | 0.04 | 15.3% |
| buildTime |  11.1s | 10.2s | -945ms | -0.85 | -9.2% |
| testBuildTime |  0ms | 0ms | 0ms | - | - |
| devPreviewResponsive |  6s | 7.7s | 1.6s | 0.73 | 21.6% |
| devManagerResponsive |  4s | 5.1s | 1.1s | 1.03 | 21.8% |
| devManagerHeaderVisible |  561ms | 886ms | 325ms | 0.85 | 36.7% |
| devManagerIndexVisible |  601ms | 931ms | 330ms | 0.83 | 35.4% |
| devStoryVisibleUncached |  911ms | 1.4s | 584ms | 0.28 | 39.1% |
| devStoryVisible |  600ms | 933ms | 333ms | 0.88 | 35.7% |
| devAutodocsVisible |  490ms | 722ms | 232ms | -0.03 | 32.1% |
| devMDXVisible |  437ms | 706ms | 269ms | -0.01 | 38.1% |
| buildManagerHeaderVisible |  475ms | 710ms | 235ms | -0.32 | 33.1% |
| buildManagerIndexVisible |  483ms | 717ms | 234ms | -0.43 | 32.6% |
| buildStoryVisible |  578ms | 747ms | 169ms | -0.52 | 22.6% |
| buildAutodocsVisible |  488ms | 686ms | 198ms | -0.14 | 28.9% |
| buildMDXVisible |  458ms | 684ms | 226ms | -0.02 | 33% |

## Greptile Summary

This PR replaces the usage of Lodash with es-toolkit across various files in the Storybook core package, aiming to improve tree-shaking and move towards ESM dependencies.

- Replaced Lodash imports with es-toolkit equivalents in multiple files, including merge, mapValues, isEqual, and debounce functions
- Updated `code/core/package.json` to remove Lodash dependency and add es-toolkit
- Modified `StoryStore.ts` to use es-toolkit functions and introduced a new `picky` function for object property selection
- Adjusted `PreviewWeb.test.ts` to use es-toolkit's merge and toMerged functions instead of Lodash
- Updated `parseArgsParam.ts` and `filterArgTypes.ts` to use es-toolkit's isPlainObject and pickBy functions respectively

### CODE REVIEW

**Review by JReinhold — COMMENTED** [review]
Looks good overall, only a few comments.

File: code/core/src/manager-api/lib/merge.ts [hunk 1]
```diff
@@ -19,5 +22,59 @@ export default <TObj = any>(a: TObj, b: Partial<TObj>) =>
...
+
+  return target as TObj;
+};
+
+export function picky<T extends Record<string, any>, K extends keyof T>(
+  obj: T,
+  keys: K[]
+): Partial<Pick<T, K>> {
+  return omitBy(pick(obj, keys), (v) => v === undefined);
+}
```
[the remark this answers is not in the document]
Comment (ndelangen) [reply code/core/src/manager-api/lib/merge.ts:80]: > can we replace this with pick from es-toolkit/compat?

I assumed that `/compat` is a stop-gap solution, and our intend would be to move away from it.

[the remark this answers is not in the document]
Comment (ndelangen) [reply code/core/src/manager-api/lib/merge.ts:80]: > Can you describe why pick from es-toolkit is inadequate?

`es-toolit` will add new keys to the object:

`pick({}, 'a')` 👉 `{ a: undefined }`

File: code/lib/codemod/package.json [hunk 2]
```diff
@@ -61,16 +61,17 @@
...
     "jscodeshift": "^0.15.1",
-    "lodash": "^4.17.21",
     "prettier": "^3.1.1",
     "recast": "^0.23.5",
     "tiny-invariant": "^1.3.1"
   },
   "devDependencies": {
     "@types/jscodeshift": "^0.11.10",
     "ansi-regex": "^6.0.1",
+    "camelcase": "^8.0.0",
```
[the remark this answers is not in the document]
Comment (ndelangen) [reply code/lib/codemod/package.json:74]: No, it strips away emojis, which we have an explicit test for to support.

**Comment by raon0211** [comment]
Hello! I’m the maintainer of es-toolkit.

We’ve fixed camelCase to better support emojis and updated the pick function to ignore nonexistent keys in the original object. Could you give our latest version, v1.21.0, a try?

File: code/core/src/core-server/utils/get-component-variable-name.ts [hunk 3]
```diff
@@ -6,6 +6,7 @@
  */
 export const getComponentVariableName = async (name: string) => {
   const camelCase = await import('camelcase');
```
Comment (JReinhold) [inline code/core/src/core-server/utils/get-component-variable-name.ts:8]: can you elaborate why this ended up being needed anyway?

File: code/core/src/manager-api/modules/layout.ts [hunk 4]
```diff
@@ -320,12 +319,15 @@ export const init: ModuleFn<SubAPI, SubState> = ({ store, provider, singleStory
...
+          ...toMerged(
+            defaultLayoutState.layout,
+            pick(options, Object.keys(defaultLayoutState.layout))
+          ),
           ...(singleStory && { navSize: 0 }),
         },
         ui: {
           ...defaultLayoutState.ui,
-          ...pick(options, Object.keys(defaultLayoutState.ui)),
+          ...toMerged(defaultLayoutState.ui, pick(options, Object.keys(defaultLayoutState.ui))),
```
Comment (JReinhold) [inline code/core/src/manager-api/modules/layout.ts:330]: I don't understand why `toMerged` is needed here?

**Review by JReinhold — APPROVED** [review]

File: code/core/src/core-server/utils/get-component-variable-name.ts (hunk 3 above)
Comment (JReinhold) [reply code/core/src/core-server/utils/get-component-variable-name.ts:8]: this just reverts back to using `camelcase` as it was before.

**Pull request merged**